parsers/mysql/mysql.go

```diff
@@ -81,16 +100,145 @@ type SlowQuery struct {
+	Query           string    `json:"query,omitempty"`
 	NormalizedQuery string    `json:"normalized_query,omitempty"`
+	DB              string    `json:"db,omitempty"`
+	HostedOn        string    `json:"hosted_on,omitempty"`
+	ReadOnly        bool      `json:"read_only,omitempty"`
+	ReplicaLag      int64     `json:"replica_lag,omitempty"`
+	Role            string    `json:"role,omitempty"`
 	skipQuery       bool
 }
 
-func (p *Parser) Init(_ interface{}) error {
+func (p *Parser) Init(options interface{}) error {
+	p.conf = *options.(*Options)
 	p.nower = &RealNower{}
+	if p.conf.Host != "" {
+		url := fmt.Sprintf("%s:%s@tcp(%s)/", p.conf.User, p.conf.Pass, p.conf.Host)
+		db, err := sql.Open("mysql", url)
+		if err != nil {
+			log.Fatal(err)
+		}
+		// run one time queries
+		p.hostedOn = getHostedOn(db)
+		p.role = getRole(db)
+
+		// update hostedOn and readOnly every <n> seconds
+		go func() {
+			defer db.Close()
+			ticker := time.NewTicker(time.Second * time.Duration(p.conf.QueryInterval))
+			for _ = range ticker.C {
+				p.readOnly = getReadOnly(db)
+				p.replicaLag = getReplicaLag(db)
+			}
+		}()
+	}
 	return nil
 }
 
+func getReadOnly(db *sql.DB) bool {
+	rows, err := db.Query("SELECT @@global.read_only;")
+	if err != nil {
+		log.Println(err)
+		return false
+	}
+	defer rows.Close()
+
+	var value bool
+
+	rows.Next()
+	if err := rows.Scan(&value); err != nil {
+		log.Println(err)
+		return false
+	}
+	if err := rows.Err(); err != nil {
+		log.Println(err)
+		return false
+	}
+	return value
+}
+
+func getHostedOn(db *sql.DB) string {
+	rows, err := db.Query("SHOW GLOBAL VARIABLES WHERE Variable_name = 'basedir';")
+	if err != nil {
+		log.Println(err)
+		return ""
+	}
+	defer rows.Close()
+
+	var varName, value string
+
+	for rows.Next() {
+		if err := rows.Scan(&varName, &value); err != nil {
+			log.Println(err)
+			return ""
+		}
+		if strings.HasPrefix(value, "/rdsdbbin/") {
+			return rdsStr
+		}
+	}
+
+	// TODO: implement ec2 detection
+
+	if err := rows.Err(); err != nil {
+		log.Println(err)
+		return ""
+	}
+	return selfStr
+}
+
+func getReplicaLag(db *sql.DB) int64 {
+	rows, err := db.Query("SHOW SLAVE STATUS")
+	if err != nil {
+		log.Println(err)
+		return -1
+	}
+	if !rows.Next() {
+		log.Println(errors.New("No slave status"))
+		return -1
+	}
+	defer rows.Close()
+
+	columns, _ := rows.Columns()
+	values := make([]interface{}, len(columns))
+	for i := range values {
+		var v sql.RawBytes
+		values[i] = &v
+	}
+
+	err = rows.Scan(values...)
+	if err != nil {
+		log.Println(err)
+		return -1
+	}
+
+	for i, name := range columns {
+		bp := values[i].(*sql.RawBytes)
+		vs := string(*bp)
+		if name == "Seconds_Behind_Master" {
+			vi, err := strconv.ParseInt(vs, 10, 64)
+			if err != nil {
+				return -1
+			}
+			return vi
+		}
+	}
+	return -1
+}
+
+func getRole(db *sql.DB) string {
+	rows, err := db.Query("SHOW MASTER STATUS")
+	if err != nil {
+		log.Println(err)
+		return "unknown"
```

Comment:
this gives me pause - the behavior for missing values/errors for getHostedOn/getReadOnly/getRole is uneven.  In the error case, we'd have ""/false/"unknown" respectively.  I wonder if all three of them should be `nil` in the error case (would obviously require some re-plumbing), and then we wouldn't include those nils in the event that we send up to honeycomb.

so instead of having "false" for the read_only column, we'd have nothing if there was an error checking it.